### PR TITLE
Streamline ObjectAdapter ASM in C#

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR simplifies the ObjectAdapter Active Servant Map (ASM) in C# by:
- removing servant locators
- simplifying default servants: they are now facet-specific (like servants), and we have now category-specific default servants and non-category specific default servants. The "" category no longer has a special wildcard meaning.
- reordering the parameters to Add. They are now in the expected/standard order, with first the key and then the value.

In addition, this PR rolls back previous updates on the master branch: the ASM is once again purely about servants, and no longer stores dispatchers (Disp).

### CODE REVIEW

File: csharp/src/Ice/ObjectAdapter.cs [hunk 1]
```diff
@@ -395,442 +430,262 @@ public void Destroy()
...
             {
-                identity = identity ?? new Identity(Guid.NewGuid().ToString(), "");
-                checkForDeactivation();
-                checkIdentity(identity.Value);
-
-                _servantManager.AddServant(dispatcher, identity.Value, facet);
-
-                return newProxy(identity.Value, proxyFactory, facet);
+                IObject? servant = null;
+                if (!_identityServantMap.TryGetValue(new IdentityPlusFacet(identity, facet), out servant))
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs]: I think in this case the non-inline declaration is clearer / cleaner. The formal type for out is IObject, not IObject?. And there is certainly a chance to get a null servant here.

File: csharp/src/Ice/ObjectAdapter.cs [hunk 2]
```diff
@@ -395,442 +430,262 @@ public void Destroy()
...
+        /// ArgumentException.</summary>
+        /// <param name="identity">The identity of the Ice object incarnated by this servant. identity.Name cannot
+        /// be empty.</param>
+        /// <param name="facet">The facet of the Ice object.</param>
+        /// <param name="servant">The servant to add.</param>
+        public void Add(Identity identity, string facet, IObject servant)
         {
-            lock (this)
+            lock (_mutex)
             {
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs]: Fixed, thanks,

File: csharp/src/Ice/ObjectAdapter.cs [hunk 3]
```diff
@@ -395,442 +430,262 @@ public void Destroy()
...
+        public IObject? Remove(Identity identity, string facet = "")
         {
-            lock (this)
+            lock (_mutex)
             {
-                checkForDeactivation();
-
-                return _servantManager.RemoveDefaultServant(category);
+                var key = new IdentityPlusFacet(identity, facet);
+                IObject? servant = null;
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/src/Ice/ObjectAdapter.cs]: See above.

File: csharp/test/Ice/adapterDeactivation/Servant.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        public IRouter.GetClientProxyReturnValue GetClientProxy(Current current) =>
+            new IRouter.GetClientProxyReturnValue(null, false);
+
+        public IObjectPrx GetServerProxy(Current current)
+        {
+            StringBuilder s = new StringBuilder("dummy:tcp -h localhost -p ");
+            s.Append(_nextPort++);
+            s.Append(" -t 30000");
+            return IObjectPrx.Parse(s.ToString(), current.Adapter.Communicator);
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/test/Ice/adapterDeactivation/Servant.cs]: Fixed.

File: csharp/test/Ice/adapterDeactivation/Servant.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        }
+
+        public IObjectPrx[] AddProxies(IObjectPrx[] proxies, Current current)
+        {
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply csharp/test/Ice/adapterDeactivation/Servant.cs]: Fixed.

File: csharp/src/Ice/ObjectAdapter.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/ObjectAdapter.cs]: what about removing the nested ifs with 
```
lock (_mutex)
{
    if (_identityServantMap.TryGetValue(new IdentityPlusFacet(identity, facet), out IObject servant) ||
        _categoryServantMap.TryGetValue(new CategoryPlusFacet(identity.Category, facet), out servant) ||
        _defaultServantMap.TryGetValue(facet, out servant))
    {
        return servant;
    }
    return null;
}
```

**Pull request merged**